### PR TITLE
Added withdraw method in ConsensusCogateway

### LINKED ISSUE #194: Implement ConsensusCogateway:withdraw()
1. Implement `ConsensusCogateway:withdraw()`
    - Declare withdraw intent
    - Transfer Utmost to the contract as an ERC20 tokens
    - Decrease the supply of tokens using `burn`
    - Call MessageBus

### PR DESCRIPTION
**Pull request by gulshanvasnani**
PR contains below changes :
1. Added `withdraw` method in ConsensusCogateway contract.
2. Added `burn` and `burnFrom` in ERC20I contract.
3. Added unit test case.

fixes #194

### CODE REVIEW

File: contracts/consensus-gateway/ConsensusCogateway.sol [hunk 1]
```diff
@@ -85,4 +94,65 @@ contract ConsensusCogateway is MasterCopyNonUpgradable, MessageBus, ConsensusGat
             _maxStorageRootItems
         );
     }
+
+    /**
+     * @notice It allows to withdraw utMOST tokens. Withdrawer needs to approve
```
Comment (deepesh-kn) [inline contracts/consensus-gateway/ConsensusCogateway.sol]:
```suggestion
     * @notice It allows withdrawing Utmost tokens. Withdrawer needs to approve
```

File: contracts/consensus-gateway/ConsensusCogateway.sol [hunk 2]
```diff
@@ -85,4 +94,65 @@ contract ConsensusCogateway is MasterCopyNonUpgradable, MessageBus, ConsensusGat
...
+     *         ConsensusCoGateway contract for amount to be withdrawn.
```
Comment (deepesh-kn) [inline contracts/consensus-gateway/ConsensusCogateway.sol]:
```suggestion
     *         consensus cogateway contract for the amount to be withdrawn.
```

File: contracts/consensus-gateway/ConsensusCogateway.sol [hunk 3]
```diff
@@ -85,4 +94,65 @@ contract ConsensusCogateway is MasterCopyNonUpgradable, MessageBus, ConsensusGat
...
+     *
+     * @dev Function requires :
+     *          - Amount must not be 0.
+     *          - Beneficiary must not be 0.
+     *          - Amount must be greater than gas price and gas limit.
```
Comment (deepesh-kn) [inline contracts/consensus-gateway/ConsensusCogateway.sol]: Please update the statement. It is incorrect.

File: contracts/consensus-gateway/ConsensusCogateway.sol [hunk 4]
```diff
@@ -85,4 +94,65 @@ contract ConsensusCogateway is MasterCopyNonUpgradable, MessageBus, ConsensusGat
...
+     *
+     * @param _amount Amount of tokens to be redeemed.
+     * @param _beneficiary The address in the origin chain where the value
+     *                     where the tokens will be released.
```
Comment (deepesh-kn) [inline contracts/consensus-gateway/ConsensusCogateway.sol]:
```suggestion
     *                     where the tokens will be withdrawn.
```

File: contracts/consensus-gateway/ConsensusCogateway.sol [hunk 5]
```diff
@@ -85,4 +94,65 @@ contract ConsensusCogateway is MasterCopyNonUpgradable, MessageBus, ConsensusGat
...
+     * @param _feeGasPrice Fee gas price at which reward will be calculated.
+     * @param _feeGasLimit Fee gas limit at which reward will be calculated.
```
Comment (deepesh-kn) [inline contracts/consensus-gateway/ConsensusCogateway.sol]:
```
     * @param _feeGasPrice Fee gas price for the reward calculation.
     * @param _feeGasLimit Fee gas limit for the reward calculation.
```

File: contracts/consensus-gateway/ConsensusCogateway.sol [hunk 6]
```diff
@@ -85,4 +94,65 @@ contract ConsensusCogateway is MasterCopyNonUpgradable, MessageBus, ConsensusGat
...
+        require(
+            _beneficiary != address(0),
+            "Beneficiary address must not be 0."
+        );
+        require(
+            _amount > _feeGasPrice.mul(_feeGasLimit),
+            "Withdrawal amount should be greater than max reward."
+        );
+
+        bytes32 hashWithdrawIntent = hashWithdrawIntent(
```
Comment (deepesh-kn) [inline contracts/consensus-gateway/ConsensusCogateway.sol]:
```suggestion
        bytes32 withdrawIntentHash = hashWithdrawIntent(
```

File: contracts/consensus-gateway/ConsensusCogateway.sol [hunk 7]
```diff
@@ -85,4 +94,65 @@ contract ConsensusCogateway is MasterCopyNonUpgradable, MessageBus, ConsensusGat
...
+        messageHash_ = MessageOutbox.declareMessage(
+            hashWithdrawIntent,
+            nonce,
+            _feeGasPrice,
+            _feeGasLimit,
+            msg.sender
+        );
+
+        require(
+            UtilityTokenInterface(address(most)).burnFrom(msg.sender, _amount),
```
Comment (deepesh-kn) [inline contracts/consensus-gateway/ConsensusCogateway.sol]: I don't think that `UtilityTokenInterface` is needed. 
`burn` and `burnFrom` are already implemented in ERC20Token, these functions must be part of `ERC20I`

File: contracts/consensus-gateway/ERC20GatewayBase.sol [hunk 8]
```diff
@@ -49,4 +53,31 @@ contract ERC20GatewayBase {
...
+     * @param _amount Amount of tokens.
+     * @param _beneficiary Beneficiary address.
+     *
+     * @return withdrawIntentHash_ Hash of withdraw intent.
+     */
+    function hashWithdrawIntent(
+        uint256 _amount,
+        address _beneficiary
+    )
+        public
```
Comment (deepesh-kn) [inline contracts/consensus-gateway/ERC20GatewayBase.sol:70]: Should this be internal ?

File: contracts/test/token/MockToken.sol [hunk 9]
```diff
@@ -116,4 +116,40 @@ contract MockToken is ERC20I, MockTokenConfig {
...
+
+    /**
+     * @notice It burns an amount of the token of a given
+     *         account.
+     * @param _account The account whose tokens will be burnt.
+     * @param _value The amount that will be burnt.
+     *
+     * @return success `true` if tokens are burned successfully otherwise false
+     */
+    function burn(address _account, uint256 _value) public returns (bool success) {
```
Comment (deepesh-kn) [inline contracts/test/token/MockToken.sol:128]: I think that `MockToken` just inherit `ERC20Token`.
I believe that there is a PR for this already. (So no need to handle this in this ticket.)

File: contracts/utility-token/UtilityTokenInterface.sol [hunk 10]
```diff
@@ -0,0 +1,51 @@
...
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+
+/**
+ * @title Utility token interface - It is an interface providing methods to
+ *        burn the tokens.
+ */
+contract UtilityTokenInterface {
```
Comment (deepesh-kn) [inline contracts/utility-token/UtilityTokenInterface.sol]: I don't think that `UtilityTokenInterface` is needed. 
`burn` and `burnFrom` are already implemented in ERC20Token, these functions must be part of `ERC20I`

File: test/consensus_cogateway/withdraw.js [hunk 11]
```diff
@@ -0,0 +1,124 @@
...
+
+  let consensusCogateway;
+  beforeEach(async () => {
+    consensusCogateway = await ConsensusCogateway.new();
+    const owner = accountProvider.get();
+    utmost = await MockToken.new(18, { from: owner });
+    param = {
+      metachainId: Utils.getRandomHash(),
+      utmostAddress: utmost.address,
+      owner,
```
Comment (deepesh-kn) [inline test/consensus_cogateway/withdraw.js]:
```suggestion
      owner: accountProvider.get(),
```
Remove `const owner = accountProvider.get();` from line 34.

**Review by deepesh-kn — CHANGES_REQUESTED** [review]

File: contracts/consensus-gateway/ConsensusCogateway.sol (hunk 1 above)
Comment (gulshanvasnani) [reply contracts/consensus-gateway/ConsensusCogateway.sol]: Done

File: contracts/consensus-gateway/ConsensusCogateway.sol (hunk 2 above)
Comment (gulshanvasnani) [reply contracts/consensus-gateway/ConsensusCogateway.sol]: Done

File: contracts/consensus-gateway/ConsensusCogateway.sol (hunk 3 above)
Comment (gulshanvasnani) [reply contracts/consensus-gateway/ConsensusCogateway.sol]: Done

File: contracts/consensus-gateway/ConsensusCogateway.sol (hunk 4 above)
Comment (gulshanvasnani) [reply contracts/consensus-gateway/ConsensusCogateway.sol]: Done

File: contracts/consensus-gateway/ConsensusCogateway.sol (hunk 5 above)
Comment (gulshanvasnani) [reply contracts/consensus-gateway/ConsensusCogateway.sol]: Done

File: contracts/consensus-gateway/ConsensusCogateway.sol (hunk 6 above)
Comment (gulshanvasnani) [reply contracts/consensus-gateway/ConsensusCogateway.sol]: Done

File: contracts/consensus-gateway/ConsensusCogateway.sol (hunk 7 above)
Comment (gulshanvasnani) [reply contracts/consensus-gateway/ConsensusCogateway.sol]: Added burn and burnFrom in ERC20I and removed `UtilityTokenInterface`.

File: contracts/consensus-gateway/ERC20GatewayBase.sol (hunk 8 above)
Comment (gulshanvasnani) [reply contracts/consensus-gateway/ERC20GatewayBase.sol:70]: `hashDepositIntent` is already `public`. Hence made `hashWithdrawIntent` `public`.

File: contracts/utility-token/UtilityTokenInterface.sol (hunk 10 above)
Comment (gulshanvasnani) [reply contracts/utility-token/UtilityTokenInterface.sol]: Added `burn` and `burnFrom` in ERC20I and removed `UtilityTokenInterface`.

File: test/consensus_cogateway/withdraw.js (hunk 11 above)
Comment (gulshanvasnani) [reply test/consensus_cogateway/withdraw.js]: Done

**Review by deepesh-kn — APPROVED** [review]

**Pull request merged**